[BE] DB Model 작업

hi0826 opened the pull request:
### 📕 Issue Number
-

<br/>

### 📙 작업 내역

> DB 모델 작업

- [x] User 모델 작업
  - User 모델 스키마, 기능 분리
    - User 스키마
      - 깃허브 login은 password가 없습니다.
        - password의 required 옵션을 해제
  - models/user 로 한 번 더 분리
  - service/auth
    - github 로그인 시에 password 정보를 같이 비교해야하므로 undefined로 넣어 추가해줬습니다.
    - github 로그인 시에 unique 값으로 준 id가 겹칠 우려가 있어 _github를 아이디 끝에 추가해줬습니다.
   
- [x] Payment 모델 작업
  - Payment 모델 스키마, 기능 분리
  - 기존의 Default Payment Method 라는 변수명이 너무 길고, 직관적이지 않은 것 같아 수정하였습니다.
  - ❗ 추후에 Account Book 이나 Transaction 에서 사용할 때, 재사용을 위해서 Schema에 다른 속성을 부여하여 재활용하는 방법도 고려중입니다.
  - ❗❗ 다른 생각하다가 커밋메세지를 잘못 작성하였습니다... 
  - ⚒ Default Payment Method Model 은 스크럼에 얘기해던대로 변경하였습니다.
  
- [x] Category 모델 작업
  - Category 모델 스키마, 기능 분리 
  - Category 역시 스키마는 Transaction과 AccountBook에 들어갑니다.
  - Category Model 자체는 기본제공 Category를 CRUD하는 용도외에 사용되지 않습니다.

- [x] Account Book 모델 작업
  - Account Book 모델 스키마, 기능 분리
  - Schema 에 default 속성을 사용하여, 값이 없는 Document도 기본으로 값을 들어가게해 데이터 일관성 유지
  - Account Book 을 새롭게 만들게되면, Default Payment Method 와 Category를 읽어와 넣을 수 있도록 기능 추가

- [x] transaction 스키마, 기능 작업
  - transaction 스키마 추가
  - Accountbook Model에서 transaction 추가기능 개발
<br/>

### 멘토님 멘셔닝

@boostcamp-2020/accountbook_mentor
<br/>

hi0826 committed:
[fix] #22 modal 화면오류 해결

- 모달창 position absolute로 화면오류 해결

hi0826 committed:
[refactor] #27 가계부 내역 컴포넌트 수정

- 가계부 내역 컴포넌트 default component 형식에 맞게 수정

hi0826 committed:
[style] transaction 컴포넌트 코드포메팅

- 내역 컴포넌트 코드포메팅

hi0826 committed:
Merge branch 'develop' of https://github.com/boostcamp-2020/Project16-C-Account-Book into transaction-interaction

hi0826 committed:
Merge branch 'develop' of https://github.com/boostcamp-2020/Project16-C-Account-Book into db-modeling

hi0826 committed:
[chore] #8 절대경로 설정, DB 연결주소 표시

- 절대경로 사용하기 쉽게 설정 추가
- DB 연결시 local / cloud DB중 어디로 연결됐는지 표시

hi0826 committed:
[feat] #8 db User 모델 schema, 모델 분리, auth service수정

- User 모델 스키마, 기능 분리
  - User 스키마
    - 깃허브 login은 password가 없음
    - password의 required 옵션을 해제
- models/user 로 한 번 더 분리
- auth
  - github 로그인 시에 password 정보를 undefined로 넣어 추가
  - github 로그인 시에 unique 값으로 준 id가 겹칠 우려가 있어 _github를 추가

hi0826 committed:
[feat] #8 category Model

- 카테고리 모델 추가
  - 변수명이 너무 길고 직관적이지 않은 것 같아서 변경
  - 카테고리 스키마와 기능 분리

j03y14 commented on BE/src/loaders/mongoose.ts:
```diff
@@ -21,7 +21,7 @@ export default async (): Promise<void> => {
         throw error;
       } else {
-        console.log('connected to DB ✅');
+        console.log(`connected to DB ${URL}✅`);
```
오 좋네요 이거 항상 어디 연결된거지 했었는데

msmk530 replied:
최종코드엔 지워야겠지만 현재로선 박수뿐...👏🏻

hi0826 replied:
여러분들이 좋아하시니 다행입니다 호호..

j03y14 commented on BE/src/models/user/schema.ts:
```diff
@@ -0,0 +1,15 @@
+export interface User extends mongoose.Document {
+  id: string;
+  name: string;
+  password: string;
```
패스워드가 들어오는 경우가 없어서 삭제하는게 좋을 것 같아요

그리고 제가 네이버 로그인까지 만들면서 `social`이라는 것까지 추가했는데 이게 좀 충돌이 있을 것 같네요ㅠㅠ

hi0826 replied:
아 생각해보니 로컬로그인이 없네요..!
지워서 반영하겠습니다.

j03y14 commented on BE/src/services/auth.ts:
```diff
@@ -43,11 +43,18 @@ const getGithubUserInfo = async (accessToken: string) => {
 };
 
-const checkUserInDB = async (userInfo: { id: string }): Promise<boolean> => {
+const checkUserInDB = async (userInfo: {
+  id: string;
+  password: string | undefined;
+}): Promise<boolean> => {
   const user = await userModel.get(userInfo);
   return !!user;
 };
```
이 함수도 제가 그냥 없애벼렸습니다. `!!(await userModel.get(userInfo))` 이렇게 하니까 되더라고요 그냥!

j03y14 commented on FE/src/components/transaction/list/transactions/transactions.tsx:
```diff
@@ -2,7 +2,12 @@
+  payment: string,
+  cost: number,
+) => {
   return (
```
이것도 어차피 컴포넌트니까 `TransactionItem` 같은 걸로 이름을 짓고 사용하는 곳에서도 
```tsx
<TransactionItem 
  category={t.category}
  description = {t.description} 
  payment={t.payment}
  cost={t.cost}
/>
```
이런식으로 하는게 좋을 것 같아요

j03y14 reviewed (approved):
홍일님 고생하셨습니다!

제가 네이버 로그인 한 부분이랑 좀 충돌이 날 것 같네요ㅠㅠ

yoonwoo123 commented:
타입스크립트를 이용해서 리턴 타입을 잘 명시해주셨네요. 감사합니다!

msmk530 commented on BE/src/models/paymentmethod/schema.ts:
```diff
@@ -0,0 +1,16 @@
+  name: { type: String, required: true },
+  color: { type: String, required: true },
+});
+
+export const PaymentModel = mongoose.model<Payment>(
+  'default_payment_methods',
+  Schema,
+);
```
나중에  default_payment_methods말고 account book 안의 payment method를 위한 모델도 만들어야할 것 같은데
이름 구분을 해두는게 좋지 않을까요 !? ... Default는 한번 만들어두고 사용될일이 없어서 구분되게 만들어도 괜찮긴 할 거 같아요 !

hi0826 replied:
default payment method  Collection은 따로 분리가 되어있는데, transaction에 들어가는 payment 정보나 accountbook에 들어가는 payment 정보는 공유되는 부분이 있더라구요. 
그래서 모델은 default로 만들되 인터페이스와 스키마를 재사용하면 좋을 것 같아서 payment로? 만들면 어떨까? 라는 생각에서 구분을 했어요!

msmk530 reviewed (approved):
default payment method랑 payment method가 interface는 같은데 모델 만들때 구분되어야 해서 
그부분 이름만 분리하는건 어떨까요 !?!?
홍일님 수고하셨습니다 ㅎㅎ

hi0826 committed:
[feat] #8 user model password 속성제거

- user model 에서 password 가 필요없어 제거
  - 소셜로그인 기능만 제공한다면, 비밀번호 필요없음

hi0826 committed:
[feat] #8 결제수단 모델 변경

- 결제수단 모델 이름 변경
  - 결제수단 스키마는 재사용할 부분
  - 결제수단 모델은 재사용하지 않을 부분이므로 이름을 변경

hi0826 committed:
[feat] #8 카테고리 모델 추가

- 카테고리 스키마, 모델 추가
  - 카테고리 스키마는 재사용
  - 카테고리 모델은 기본 카테고리를 읽는용도 외에 사용X

hi0826 committed:
[feat] #8 transaction 스키마

- transaction 은 직접 DB에 접근할 일이 없습니다.
  - AccountBook에 들어갈 요소 이므로, 코드의 가독성을 높이고자 스키마로 분리
  - 따라서 스키마와 interface 정도만 공유되면 될 것 같습니다.

hi0826 committed:
[feat] #8 user 스키마 변경

- password 속성 필요없어 삭제

hi0826 committed:
[feat] #8 payment method 스키마 변경

- 아무 값이 없어도 속성을 유지할 수 있게 default 속성 추가

hi0826 committed:
[fix] #8 category 모델 변경

- 카테고리 Colection 명 복수로 안해서 제대로 생성안되어 변경

hi0826 committed:
[feat] #8 account book 모델 추가, 서비스 기능변경

- account book 스키마 생성
  - 없어도 되는 부분들은 default 속성으로 기본 값이 들어가게 변경
- account book 기능 생성
  - account book 모델의 기능 생성
    - 기본으로 생성시 default 로 들어가는 payment와 category를 넣도록 추가
- account book 서비스 변경
  - 스키마와 기능의 변경으로 약간의 수정

hi0826 committed:
[feat] #8 accountbook 스키마 수정

- Document 타입 정의시 잘못된 부분 수정

hi0826 committed:
[feat] #8 transaction 스키마 추가

- 거래내역 스키마 및 모델 추가
- 모델은 사용하지 않으므로 추후 제거 가능성 있음

hi0826 committed:
[feat] #8 accountbook service 변경, 개발

- accountbook 모델 함수 변경으로 코드수정
- accountbook model import 시 좀 더 직관적으로 변수명 변경
- 거래내역 추가시 기능개발할 수 있도록 service 추가

hi0826 committed:
[feat] #8 거래내역 추가 기능

- 거래내역 추가시 가계부에 들어가도록 설정
- parameter, return 타입, 맞춰서 변경
- 가계부에서 transaction을 찾아와서 업데이트 시키고 accountbook model을 업데이트 시키는 방식

j03y14 commented on BE/src/models/accountbook/index.ts:
```diff
@@ -0,0 +1,50 @@
+    transactions: [],
+  };
+
+  let result = await new AccountBookModel(accountbook);
+  result = await result.save();
```
`result`를 return 안 하면 
```typescript
  const result = await new AccountBookModel(accountbook);
  await result.save();
```
이렇게 해도 될 것 같아요!

msmk530 replied:
accountbook을 생성하고 나서 해당 _id값을 반환해주어야 할 것 같은데 result에 들어가는 건가요 !?!? 
result에 들어가는 거면 return해서 front로 _id값 주면 될 것 같아요 !!

hi0826 replied:
result에는 생성된 도큐먼트가 반환되더라구요.
리팩터링으로 남겨둔 부분인데 오늘 API 짜면서 필요한 부분 꺼내서 리턴하면 될 것 같습니다~

j03y14 commented on BE/src/models/accountbook/index.ts:
```diff
@@ -0,0 +1,50 @@
+  const id = accountBookId;
+  const curAccountBook = await AccountBookModel.findOne({ _id: id });
+  if (curAccountBook) {
+    const curTransactions = [...curAccountBook.transactions];
+    curTransactions.push(transaction);
```
```typescript
const curTransactions = [...curAccountBook.transactions, transaction];
```
이렇게 할 수도 있을것 같아요!

j03y14 reviewed (approved):
수고하셨습니다 홍일님!👍

msmk530 reviewed (approved):
고생하셨습니다 ㅎㅎㅎ 좋은거 같아요 !!

hi0826 committed:
Merge branch 'develop' into db-modeling

hi0826 committed:
[feat] #8 Account Book Collection 변경

- Account Book
  - users 에 있는 id 들의 배열을 user 배열로 바꿔줌

hi0826 committed:
Merge branch 'db-modeling' of https://github.com/boostcamp-2020/Project16-C-Account-Book into db-modeling
